southpark-incremental-indexing/README.md

```diff
@@ -0,0 +1,163 @@
+
+Run the indexing step
+
+`python app.py -t index`
+
+This will perform the indexing on the first file, stop the flow, re-start it, and then index the second. This showcases **incremental indexing**.
+
+Query the data
+
+`python app.py -t query`
+
+
+## Overview
+
+For the discussion on the basic topics, read through the documentation in the [South Park example](../southpark-search/README.md). This README will delve deeper into how we use **incremental indexing**.
+
+### What is incremental indexing?
+
+Briefly, this allows for the user to re-use an index to add new data. It also automatically handles adding documents with duplicate IDs. For a more in-depth explanation on how Jina handles this, refer to [this chapter](https://docs.jina.ai/chapters/prevent_duplicate_indexing/index.html) in our docs.
+
+### Configuration changes
+
+In order to adapt the South Park scripts example to support incremental indexing, we need to to the following:
+
+1. Change the `indexer` in `flow-index` and `flow-query` to use `DocIDCache` as a filter (in the `uses_before` field). This ensures that we prevent duplicates.
+
+    In this example the `DocIDCache` is separated into its own `.yml` file, in `index_cache`:
+    
+    ```yaml
+    !DocIDCache
+    with:
+      index_path: cache.tmp
+    metas:
+      name: cache
+      workspace: $JINA_WORKSPACE
+    requests:
+      on:
+        [SearchRequest, TrainRequest, IndexRequest, ControlRequest]:
+          - !RouteDriver {}
+        IndexRequest:
+          - !TaggingCacheDriver
+            with:
+              tags:
+                is_indexed: true
+          - !FilterQL
+            with:
+              lookups: {tags__is_indexed__neq: true}
+    ```
+    
+    This might look complicated, but what it basically does is to first check the cache for any matching doc IDs, before indexing and querying.
+
+2. Adapt dataset to highlight incremental indexing.
+
+    We split the dataset into two files, `character-lines-1.csv` and `characters-lines-2.csv`. This way we can show how we can index one, close the `Flow` object, and then index the other.
+    
+    The env. var. `JINA_DATA_FILE` has also been split into two, with `JINA_DATA_FILE_1` and `JINA_DATA_FILE_2`. By default these are set to point to the two files above.
+
+3. Adapt the `app.py` to index, close, and index again.
+
+    When running `python app.py -t index` we would usually only index one file. We now have the following:
+    
+    ```python
+    f = Flow().load_config("flow-index.yml")
+
+    with f:
+        print(f'Indexing file {os.environ["JINA_DATA_FILE_1"]}')
+        f.index_lines(
+        ...
+        )
+
+    # we then re-use the same index to append new data
+    with f:
+        print(f'Indexing file {os.environ["JINA_DATA_FILE_2"]}')
+        f.index_lines(
+           ...
+        )    
+    ```
+    
+    This indexes the first file, closes the flow, and then indexes a second file.
```

Comment:
Is it possible to show an example of query showing that it returns documents from both parts?

Reply:
It's in the integration test now